tools/scripts/add_library_tso_doublets.py

```diff
@@ -0,0 +1,180 @@
+    doublet_obs_names = ["X" + str(i) for i in range(n_doublets)]
+    doublet_adata = ad.AnnData(X=doublet_X, obs=pd.DataFrame(index=doublet_obs_names), var=pd.DataFrame(index=adata.var_names))    
+    adata = adata.concatenate(doublet_adata, index_unique=None)
+
+    adata.obs["doublet_cell"] = adata.obs_names.isin(doublet_obs_names)
+    adata.obs["doublet_cell"] = adata.obs["doublet_cell"].astype("category")
+    adata.layers["UMIs"] = adata.X.copy()
+    sc.pp.normalize_total(adata, target_sum=1e4)
+    sc.pp.log1p(adata)
+
+    try:
+        sc.pp.highly_variable_genes(adata, n_top_genes=5000, flavor="seurat_v3", layer="UMIs")
+        adata.layers["UMIs"]
+
+    except:
+        sc.pp.highly_variable_genes(adata, min_mean=1, min_disp=0.5)
+        del adata.layers["UMIs"]
+
+    adata_sub = adata[:, adata.var["highly_variable"]].copy()
+    sc.pp.scale(adata_sub)
+    sc.pp.pca(adata_sub)
+    v = adata_sub.uns['pca']['variance']
+    n_pcs = np.max(np.where(((v - np.mean(v)) / np.std(v)) > 0)[0])   
+    knn = KNeighborsTransformer(
+        n_neighbors=k,
+        algorithm="kd_tree",
+        n_jobs=-1,
+    )
+
+    knn = knn.fit(adata_sub.obsm["X_pca"][:, :n_pcs])
+    dist, idx = knn.kneighbors()  
+    knn_mapper = KNeighborsTransformer(
+        n_neighbors=10,
+        algorithm="kd_tree",
+        n_jobs=-1,
+    )
+
+    knn_mapper = knn_mapper.fit(adata_sub[adata_sub.obs["doublet_cell"] == False, :].obsm["X_pca"][:, :n_pcs])
+    dist1, _ = knn_mapper.kneighbors(adata_sub[adata_sub.obs["doublet_cell"] == True, :].obsm["X_pca"][:, :n_pcs])    
+    dist_th = np.mean(dist1) + (1.64 * np.std(dist1))
+    freq = (dist < dist_th) & (idx > adata[adata.obs["doublet_cell"] == False, :].shape[0])
+    score1 = freq.mean(axis=1)
+    score2 = freq[:, :np.int64(np.ceil(k/2))].mean(axis=1)
+    adata.obs["doublet_score"] = [np.max([score1[i], score2[i]]) for i in range(adata.shape[0])]   
+    doublet_csv=adata.obs.loc[~adata.obs_names.isin(doublet_obs_names), ["doublet_score"]]
+    
+    # Calculate the percentage of doublets with a doublet_score > 0.3
+    num_doublets = doublet_csv[doublet_csv["doublet_score"] > 0.3].shape[0]
+    total_cells = doublet_csv.shape[0]
+    percent_doublets = num_doublets / total_cells * 100
+
+    return doublet_csv, percent_doublets 
+
+
+# Function to calculate additional library metrics such as keeper cells
+def process_gex_data(gex_h5ad_modified, gex_nhash_id, library_csv, input_id, doublets, doublet_scores, counting_mode, expected_cells):
+    print("Reading Optimus h5ad:")
+    gex_data = gex_h5ad_modified
+    # NHashID is optional input, so the logic below sets it if undefined
+    if gex_nhash_id is not None:
+        gex_data.uns['NHashID'] = gex_nhash_id
+    else:
+        gex_nhash_id = "NA"
+        gex_data.uns['NHashID'] = gex_nhash_id
+
+    #gex_data.write(f"{input_id}.h5ad")
+
+    print("Reading library metrics")
+    library = pd.read_csv(library_csv, header=None)
```

Comment:
pandas DataFrame handling can be improved here:
   library = pd.read_csv(library_csv, header=None, index_col=0, squeeze=True)

Reply:
could not use squeeze as is